### ISSUE TITLE
关于mxnet内存使用的一些问题

### ISSUE THREAD

**Issue by albertyou2** [dmlc/mxnet#2682]
我的环境是 mac 10.11.4 ，mxnet 0.7.0 ，使用openmp 和kv，CPU训练。这段时间一直使用mxnet做cnn训练。
1. 问题一：使用内存和mxnet-memonger 探测的差距很大

我原来的train 数据集是 200 ＊200 ＊3的图像，1076张,batch_size ＝ 20，我使用如下代码提供数据：
train = mx.io.ImageRecordIter(
        #path_imgrec = os.path.join(args.data_dir, args.train_dataset),
        path_imgrec   = train_file,
        mean_img    = train_min,
        shuffle     = True,
        data_shape  = data_shape,
        batch_size  = args.batch_size,
        rand_crop   = True,
        rand_mirror = True,
        num_parts   = kv.num_workers,
        part_index  = kv.rank)
使用例子中的vgg，train_model.py,train_imagenet.py训练
我发现会消耗掉 5.1G 左右的内存。我使用mxnet-memonger探测，只需要744MB.是我那里配置错误么？windows下同样存在类似问题

问题二：mxnet 是否一次性加载.rec文件进入内存？
我的数据集换成70000万张图，train.rec文件300mb，训练模型文件等一切同上，我发现内存使用达到7.9g。

按照我的理解，内存使用应该变对么？因为我的shape,和bath_size，网络都没有变化

问题三：入股大文件会导致内存使用过多，那么我该如何将 原始文件分别处理维rec，以及如何喂给模型呢？
谢谢！！！

**Comment by piiswrong** [comment]
memonger is not activated by default. You need to call it explicitly.

2. This is due to preloading. I think there is a parameter you can tune to use less memory @tqchen

**Comment by albertyou2** [comment]
@piiswrong 谢谢回复！ 

   是的，我显式调用了mxnet-memonger ，使用如下代码：
    ......
    import memonger
    ......
    dshape = (args.batch_size, 3, 200, 200)
    net_mem_planned = memonger.search_plan(network, data=dshape)
    old_cost = memonger.get_cost(network, data=dshape)
    new_cost = memonger.get_cost(net_mem_planned, data=dshape)

    然后我训练使用的是 ：
    model = mx.model.FeedForward(
    ctx                = devs,
    symbol             = net_mem_planned,
    num_epoch          = args.num_epochs,
    learning_rate      = args.lr,
    momentum           = 0.9,
    wd                 = 0.00001,
    initializer        = mx.init.Xavier(factor_type="in", magnitude=2.34),
    **model_args
    )
    训练之前有如下打印

**Comment by piiswrong** [comment]
@tqchen

**Comment by Godricly** [comment]
check this issue #2087.
The memory returned by get_cost only considers part of actual cost.

**Comment by albertyou2** [comment]
@Godricly  thanks !
